glotaran/utils/json_schema.py

```diff
@@ -0,0 +1,121 @@
+
+    Since the actual GlotaranDataModel is created dynamically we create a schema that has the
+    attributes all subclasses.
+
+    We do this in a cached function to:
+    - Save time recreating the schema (usage with file watcher to update parameters)
+    - Not recreate ``data_model_class`` since this will lead to issues when run in an interactive
+        session since each ``GlotaranDataModel`` class is a subclass of :class:`DataModel`.
+
+    Returns
+    -------
+    tuple[dict[str, Any], dict[str, Any]]
+        Json Schema for :class:`Scheme` and a :class:`DataModel`
+    """
+    data_model_class = create_model(
+        "GlotaranDataModel", __base__=tuple(DataModel.__subclasses__())
+    )
+    return Scheme.model_json_schema(), data_model_class.model_json_schema()
+
+
+def _create_vanilla_schema() -> tuple[dict[str, Any], dict[str, Any]]:
+    """Extra layer since mutable objects and cached return values don't play nicely.
+
+    Returns
+    -------
+    tuple[dict[str, Any], dict[str, Any]]
+        Copy of cached Json Schema from ``_create_vanilla_schema_cached``.
+    """
+    json_schema, data_model_schema = _create_vanilla_schema_cached()
+    return deepcopy(json_schema), deepcopy(data_model_schema)
+
+
+@overload
+def create_model_scheme_json_schema(
+    output_path: None = None, parameters: Parameters | StrOrPath | None = None
+) -> dict[str, Any]:
+    ...
+
+
+@overload
+def create_model_scheme_json_schema(
+    output_path: StrOrPath, parameters: Parameters | StrOrPath | None = None
+) -> None:
+    ...
+
+
+def create_model_scheme_json_schema(
+    output_path: StrOrPath | None = None, parameters: Parameters | StrOrPath | None = None
+) -> dict[str, Any] | None:
+    """Create the json scheme for the model scheme.
+
+    Parameters
+    ----------
+    output_path : StrOrPath | None
+        Path to write the schema to. Defaults to None
+    parameters : Parameters | StrOrPath | None
+        Parameters to inject labels into schema. Defaults to None
+
+    Returns
+    -------
+    dict[str, Any] | None
+        Json Schema dict if no ``output_path`` was provided.
+    """
+    json_schema, data_model_schema = _create_vanilla_schema()
+
+    # Overwrite required fields with required fields from base class
+    data_model_schema["required"] = [
+        name for name, field in DataModel.model_fields.items() if field.is_required()
+    ]
+    json_schema["$defs"] |= data_model_schema.pop("$defs")
+    json_schema["$defs"]["GlotaranDataModel"] = data_model_schema
+    json_schema["$defs"]["ExperimentModel"]["properties"]["datasets"]["additionalProperties"][
+        "$ref"
+    ] = "#/$defs/GlotaranDataModel"
+
+    # We need this overwrite since we generate json schema for editor support (auto completion and
+    # linting) of model schema files where the parameter label is used rather than defining a
+    # parameter.
+    parameter_label_schema: dict[str, str | list[str]] = {
```

Comment:
Another question: why not make this parameter schema things part of the `ParameterType` alias?